### ISSUE TITLE
rotation invariant similarity duplicates

### ISSUE THREAD

**Issue by mowgli** [BestImageViewer/geeqie#347]
The patch "compare.diff" modifies the functions image\_sim\_compare and image\_sim\_compare\_fast so that it compares with the eight possible isometric transformations \(compositions of 90°-rotations, mirrors, transpose,...\), using the same similarity signature.

The patch "fill\_data.diff" is a slight modification of the image\_sim\_fill\_data, that creates the similarity signature. When the image size is not a multiple of 32, all cells had size w/32, so the last w%32 pixels were just ignored, and more importantly it was not symmetric at all when the signature was built from the other side, which is annoying for my comparison algorithm. So I changed the cell size \(x\_inc\) from a constant one to "round\(w\_left/n\_left\)", in order to spread the spare pixels \(eg to split 42 pixels into 4 cells, it results in 11,10,11,10 instead of 10,10,10,10\). With an image of size w with w%32=16, my rotation-invariant algorithm results in a similarity of 0.951000 with the original version, and 0.999968 with this modification \(apparently lossless jpeg rotation is not perfectly lossless\).

I haven't done any special integration, I just modify the default behavior, so that you can test it and decide how to integrate it, maybe this should be an option. I haven't modified the "alternate" comparison algorithm, but it is straightforward as my algorithm only changes which cells it compares, not how it compares them.

Any comment, any idea, any question, feel free :-\)

Reported by: cyril42e

**Comment by mowgli** [comment]
Original patch:
```patch
Index: src/similar.c
===================================================================
--- src/similar.c	(revision 1915)
+++ src/similar.c	(working copy)
@@ -311,51 +311,100 @@
-gdouble image_sim_compare(ImageSimilarityData *a, ImageSimilarityData *b)
-{
-	gint sim;
-	gint i;
-
-	if (!a || !b || !a->filled || !b->filled) return 0.0;
-
-	sim = 0.0;
-
-	for (i = 0; i < 1024; i++)
-		{
-		sim += abs(a->avg_r[i] - b->avg_r[i]);
-		sim += abs(a->avg_g[i] - b->avg_g[i]);
-		sim += abs(a->avg_b[i] - b->avg_b[i]);
-		}
-
-	return 1.0 - ((gdouble)sim / (255.0 * 1024.0 * 3.0));
-}
-
-/* this uses a cutoff point so that it can abort early when it gets to
- * a point that can simply no longer make the cut-off point.
- */
-gdouble image_sim_compare_fast(ImageSimilarityData *a, ImageSimilarityData *b, gdouble min)
-{
-	gint sim;
-	gint i;
-	gint j;
-
-#ifdef ALTERNATE_INCLUDE_COMPARE_CHANGE
-	if (alternate_enabled) return alternate_image_sim_compare_fast(a, b, min);
-#endif
-
-	if (!a || !b || !a->filled || !b->filled) return 0.0;
-
-	min = 1.0 - min;
-	sim = 0.0;
-
-	for (j = 0; j < 1024; j += 32)
-		{
-		for (i = j; i < j + 32; i++)
-			{
-			sim += abs(a->avg_r[i] - b->avg_r[i]);
-			sim += abs(a->avg_g[i] - b->avg_g[i]);
-			sim += abs(a->avg_b[i] - b->avg_b[i]);
-			}
-		/* check for abort, if so return 0.0 */
-		if ((gdouble)sim / (255.0 * 1024.0 * 3.0) > min) return 0.0;
-		}
-
-	return (1.0 - ((gdouble)sim / (255.0 * 1024.0 * 3.0)) );
-}
+gdouble image_sim_compare_transfo(ImageSimilarityData *a, ImageSimilarityData *b, gchar transfo)
+{
+	gint sim;
+	gint i1, i2, *i;
+	gint j1, j2, *j;
+
+	if (!a || !b || !a->filled || !b->filled) return 0.0;
+
+	sim = 0.0;
+
+	if (transfo & 1) { i = &j2; j = &i2; } else { i = &i2; j = &j2; }
+	for (j1 = 0; j1 < 32; j1++)
+		{
+		if (transfo & 2) *j = 31-j1; else *j = j1;
+		for (i1 = 0; i1 < 32; i1++)
+			{
+			if (transfo & 4) *i = 31-i1; else *i = i1;
+			sim += abs(a->avg_r[i1*32+j1] - b->avg_r[i2*32+j2]);
+			sim += abs(a->avg_g[i1*32+j1] - b->avg_g[i2*32+j2]);
+			sim += abs(a->avg_b[i1*32+j1] - b->avg_b[i2*32+j2]);
+			}
+		}
+
+	return 1.0 - ((gdouble)sim / (255.0 * 1024.0 * 3.0));
+}
+
+gdouble image_sim_compare(ImageSimilarityData *a, ImageSimilarityData *b)
+{
+	gboolean test_transformations = TRUE; /* could be a function parameter */
+	gint max_t = (test_transformations ? 8 : 1);
+	
+	gint t;
+	gdouble score, max_score = 0;
+	
+	for(t = 0; t < max_t; t++) 
+	{
+		score = image_sim_compare_transfo(a, b, t);
+		if (score > max_score) max_score = score;
+	}
+	return max_score;
+}
+
+
+/*
+4 rotations (0, 90, 180, 270) combined with two mirrors (0, H)
+generate all possible isometric transformations
+= 8 tests
+= change dir of x, change dir of y, exchange x and y = 2^3 = 8
+*/
+gdouble image_sim_compare_fast_transfo(ImageSimilarityData *a, ImageSimilarityData *b, gdouble min, gchar transfo)
+{
+	gint sim;
+	gint i1, i2, *i;
+	gint j1, j2, *j;
+
+#ifdef ALTERNATE_INCLUDE_COMPARE_CHANGE
+	if (alternate_enabled) return alternate_image_sim_compare_fast(a, b, min);
+#endif
+
+	if (!a || !b || !a->filled || !b->filled) return 0.0;
+
+	min = 1.0 - min;
+	sim = 0.0;
+
+	if (transfo & 1) { i = &j2; j = &i2; } else { i = &i2; j = &j2; }
+	for (j1 = 0; j1 < 32; j1++)
+		{
+		if (transfo & 2) *j = 31-j1; else *j = j1;
+		for (i1 = 0; i1 < 32; i1++)
+			{
+			if (transfo & 4) *i = 31-i1; else *i = i1;
+			sim += abs(a->avg_r[i1*32+j1] - b->avg_r[i2*32+j2]);
+			sim += abs(a->avg_g[i1*32+j1] - b->avg_g[i2*32+j2]);
+			sim += abs(a->avg_b[i1*32+j1] - b->avg_b[i2*32+j2]);
+			}
+		/* check for abort, if so return 0.0 */
+		if ((gdouble)sim / (255.0 * 1024.0 * 3.0) > min) return 0.0;
+		}
+
+	return (1.0 - ((gdouble)sim / (255.0 * 1024.0 * 3.0)) );
+}
+
+/* this uses a cutoff point so that it can abort early when it gets to
+ * a point that can simply no longer make the cut-off point.
+ */
+gdouble image_sim_compare_fast(ImageSimilarityData *a, ImageSimilarityData *b, gdouble min)
+{
+	gboolean test_transformations = TRUE; /* could be a function parameter */
+	gint max_t = (test_transformations ? 8 : 1);
+	
+	gint t;
+	gdouble score, max_score = 0;
+	
+	for(t = 0; t < max_t; t++) 
+	{
+		score = image_sim_compare_fast_transfo(a, b, min, t);
+		if (score > max_score) max_score = score;
+	}
+	return max_score;
+}


```

image_sim_fill_data patch:
```patch
Index: src/similar.c
===================================================================
--- src/similar.c	(revision 1915)
+++ src/similar.c	(working copy)
@@ -169,6 +169,13 @@
 #endif
 }
 
+gint mround(gdouble x) 
+{ 
+	gint ipart = x;
+	gdouble fpart = x-ipart; 
+	return (fpart < 0.5 ? ipart : ipart+1); 
+}
+
 void image_sim_fill_data(ImageSimilarityData *sd, GdkPixbuf *pixbuf)
 {
 	gint w, h;
@@ -182,10 +189,10 @@
 	gint j;
 	gint x_inc, y_inc, xy_inc;
 	gint xs, ys;
+	gint w_left, h_left;
 
 	gboolean x_small = FALSE;	/* if less than 32 w or h, set TRUE */
 	gboolean y_small = FALSE;
-
 	if (!sd || !pixbuf) return;
 
 	w = gdk_pixbuf_get_width(pixbuf);
@@ -197,6 +204,8 @@
 	p_step = has_alpha ? 4 : 3;
 	x_inc = w / 32;
 	y_inc = h / 32;
+	w_left = w;
+	h_left = h;
 
 	if (x_inc < 1)
 		{
@@ -209,16 +218,16 @@
 		y_small = TRUE;
 		}
 
-	xy_inc = x_inc * y_inc;
-
 	j = 0;
 
+	h_left = h;
 	for (ys = 0; ys < 32; ys++)
 		{
 		if (y_small) j = (gdouble)h / 32 * ys;
-
+		        else y_inc = mround((gdouble)h_left/(32-ys));
 		i = 0;
 
+		w_left = w;
 		for (xs = 0; xs < 32; xs++)
 			{
 			gint x, y;
@@ -227,7 +236,8 @@
 			guchar *xpos;
 
 			if (x_small) i = (gdouble)w / 32 * xs;
-
+			        else x_inc = mround((gdouble)w_left/(32-xs));
+			xy_inc = x_inc * y_inc;
 			r = g = b = 0;
 			xpos = pix + (i * p_step);
 
@@ -253,9 +263,11 @@
 			sd->avg_b[t] = b;
 
 			i += x_inc;
+			w_left -= x_inc;
 			}
 
 		j += y_inc;
+		h_left -= y_inc;
 		}
 
 	sd->filled = TRUE;
```

**Comment by mowgli** [comment]
@cyril42e, maybe that is your request from Sourceforge?

**Comment by mowgli** [comment]
Implemented with merge ce4ed47.

Thanks to Cyril Roussillon (@cyril42e?) for this great idea.

**Comment by cyril42e** [comment]
Hi @mowgli, yes it was my request on Sourceforge, nice to see this finally merged, thanks!
